### DISCUSSION TITLE
VLAN Support

### DISCUSSION THREAD

**Discussion by mimik13** [humbertogontijo/homeassistant-roborock#96] [General]
Hi. Does this integration require the vacuum to be on the same VLAN as Home Assistant or will it work across VLANs?

Thanks.

**Comment by jdejaegh** [comment]
Not tested across VLAN but it works across networks: vacuum connected to a 4G mobile hotspot and HA on my DSL residential connection.
So this should work across VLAN as well.

As the integration relies on Roborock's cloud it should work as long as both HA and the vacuum can access the cloud, no matter if the are on the same network.

**Comment by bobzer** [comment]
Would it be possible to not rely on the cloud ?
local control would be great

**Reply by humbertogontijo** [reply]
I don't know yet how to do it. I know that Roborock uses tuya protocol, but do not expose port 6668 like any other tuya device

**Reply by StaleLoafOfBread** [reply]
For what it's worth, the android app appears to have some local capabilities. I unplugged my modem from my router and was able to still adjust the settings on my S7 ultra. I know it worked as the robot beeped. I did not try starting a vacuum session as it's late here and people are sleeping. 

All this to say, I think there's hope for local control.

**Reply by StaleLoafOfBread** [reply]
I was not able to see any traffic via fiddler when adjusting settings. I don't have the right set up to capture traffic otherwise. Hopefully someone else can take a look to see what's going on here.

**Reply by jdejaegh** [reply]
The robot is locally broadcasting some data over UDP every 5 seconds or so.  Here is what it looks like in Wireshark from my laptop.

<details>
  <summary>Network capture</summary>
 
  ```
  Frame 115: 121 bytes on wire (968 bits), 121 bytes captured (968 bits) on interface \Device\NPF_{93C6F631-8590-4D0E-ADCA-466509D492A4}, id 0
    Section number: 1
    Interface id: 0 (\Device\NPF_{93C6F631-8590-4D0E-ADCA-466509D492A4})
    Encapsulation type: Ethernet (1)
    Arrival Time: Jan  9, 2023 19:51:56.238974000 Paris, Madrid
    [Time shift for this packet: 0.000000000 seconds]
    Epoch Time: 1673290316.238974000 seconds
    [Time delta from previous captured frame: 0.500144000 seconds]
    [Time delta from previous displayed frame: 24.986688000 seconds]
    [Time since reference or first frame: 26.723092000 seconds]
    Frame Number: 115
    Frame Length: 121 bytes (968 bits)
    Capture Length: 121 bytes (968 bits)
    [Frame is marked: False]
    [Frame is ignored: False]
    [Protocols in frame: eth:ethertype:ip:udp:data]
    [Coloring Rule Name: UDP]
    [Coloring Rule String: udp]
Ethernet II, Src: BeijingR_3b:5f:7a (b0:4a:39:3b:5f:7a), Dst: Broadcast (ff:ff:ff:ff:ff:ff)
    Destination: Broadcast (ff:ff:ff:ff:ff:ff)
    Source: BeijingR_3b:5f:7a (b0:4a:39:3b:5f:7a)
    Type: IPv4 (0x0800)
Internet Protocol Version 4, Src: 192.168.42.24, Dst: 192.168.42.255
    0100 .... = Version: 4
    .... 0101 = Header Length: 20 bytes (5)
    Differentiated Services Field: 0x00 (DSCP: CS0, ECN: Not-ECT)
    Total Length: 107
    Identification: 0x0000 (0)
    010. .... = Flags: 0x2, Don't fragment
    ...0 0000 0000 0000 = Fragment Offset: 0
    Time to Live: 64
    Protocol: UDP (17)
    Header Checksum: 0x641a [validation disabled]
    [Header checksum status: Unverified]
    Source Address: 192.168.42.24
    Destination Address: 192.168.42.255
User Datagram Protocol, Src Port: 38145, Dst Port: 58866
    Source Port: 38145
    Destination Port: 58866
    Length: 87
    Checksum: 0x8b6c [unverified]
    [Checksum Status: Unverified]
    [Stream index: 5]
    [Timestamps]
    UDP payload (79 bytes)
Data (79 bytes)
    Data: 312e300000000100020040c3850de14011b6eeb723971516ac15329f2e3d92f96654b87c…
    [Length: 79]
  ```
</details>

Using [PCAPdroid](https://f-droid.org/packages/com.emanuelef.remote_capture/) I was able to see once a TCP connection between my phone and the robot locally.  I was not recording the PCAP file and I can't get it to do it again.  Blocking DNS or removing Internet access prevents the app from opening.

**Reply by jdejaegh** [reply]
I managed to make the app connect locally to the robot again.  I disabled DNS resolution for the robot (using Pi-Hole and setting all address as blocked) but left my phone with full Internet access.  I captured the traffic using [PCAPdroid](https://f-droid.org/packages/com.emanuelef.remote_capture/).

The app connects to the robot on port 58867/tcp (which is discovered as open by nmap).  They then exchange gibberish (i.e. the data does not translate to plain ASCII text) over that TCP connection.  The data does not seem to be encrypted as the payloads seem very similar to each other and have a low entropy.

Here is an example of the first payload sent by the app to the robot

```
0000   00 00 00 15 31 2e 30 00 00 00 00 00 00 00 00 00   ....1.0.........
0010   00 00 00 00 00 00 00 00 1e                        .........
```

I now have a PCAP dump of the TCP connection that happened between the app and the robot with me changing mop and vacuum settings.  I am not sure that everything between the robot and the app went through that TCP connection as the robot may have some DNS cache and may have kept a connection to Roborock's servers.  During the recording, the app kept an active connection to Roborock MQTT server and was constantly sending data.

I don't have time to reverse engineer the network protocol but I can share the PCAP file of that local TCP connection.  The file contains 1550 packets and spans over 472 seconds.

**Reply by StaleLoafOfBread** [reply]
I see you got it working again but FYI my findings show that in order for the app to open it needs internet. Once you "enter" the robot, you can unplug your modem (or fw rules) and it will communicate locally. I can't recall if this only works for a limited amount of time or not.

**Reply by humbertogontijo** [reply]
I've tried using tuya protocol directly in port 58867 using device duid and local key, but, unfortunately, had no success

**Reply by shred86** [reply]
Has anyone tried [this](https://github.com/rytilahti/python-miio/issues/1107#issuecomment-997015096) to see if they're able to obtain any more information on how the Roborock app connects to the device locally? I suppose you would need to setup the app in an iOS or Android development environment on a computer to capture that main.bundle file.

**Comment by humbertogontijo** [comment]
I managed to execute a command offline! I'll be adding a offline.py file to repository https://github.com/humbertogontijo/python-roborock

**Reply by Lash-L** [reply]
> I managed to execute a command offline! I'll be adding a offline.py file to repository https://github.com/humbertogontijo/python-roborock

Oh dang!!! Great work - I'll check it out later and look I'm super curious. Let me know how I can help

**Reply by StaleLoafOfBread** [reply]
That's awesome! I'm super interested in seeing how you got it done!

**Reply by humbertogontijo** [reply]
It's there. A simple example, kind of messed up, just so u guys can see. The example Is "get_status", which is the one I'm trying to get the response. But you can change to "app_start", for example, and will see the vacuum starts

**Reply by Lash-L** [reply]
@humbertogontijo Tested it, and it doesn't seem to work for me. The local key is the key from HomeDataDevice, right?


`  File "E:/Documents/GitHub/python-roborock/offline.py", line 118, in _decode_msg
    decrypted_payload = unpad(decipher.decrypt(payload), AES.block_size)
  File "E:\Documents\GitHub\python-roborock\sh\lib\site-packages\Crypto\Util\Padding.py", line 92, in unpad
    raise ValueError("Padding is incorrect.")
ValueError: Padding is incorrect.
`

Any guesses? Want me to send you my payload? I have a S7

**Reply by Lash-L** [reply]
Ah - Removing the for msg in msgs: works for me. If I instead just keep the 
```
response = send_command(local_ip, RoborockCommand.GET_STATUS)
print(response)
```

I get 
`{'version': b'1.0', 'timestamp': 1680183249, 'protocol': 5}`

**Reply by humbertogontijo** [reply]
Made some changes to it. I only need get_map_v1 to work now. Anyone able to help trying figuring this out?

**Reply by humbertogontijo** [reply]
BTW, if anyone is interested, I used https://www.intuitibits.com/products/airtool2/ and Wireshark to get to this point

**Reply by Lash-L** [reply]
@humbertogontijo I'd be happy to help, but it seems as if it is not working for me.

The ip was hardcoded in socket_listener, so I changed that. 

But then, I get an unexpeccted crc32 error. my crc32 is 1188946763. If I comment out hte crc32 check, it will run, but I end up getting this output:

```
Requesting with prefix 199 and payload b'{"dps":{"101":"{\\"id\\":10001,\\"method\\":\\"get_status\\",\\"params\\":[],\\"security\\":{\\"endpoint\\":\\"abc\\",\\"nonce\\":\\"EA75AC5ABDF816F8ACAB00C8FA334444\\"}}"},"t":1680194864}'
Response with prefix 17 and payload {'version': b'1.0', 'timestamp': 1680194866, 'protocol': 5, 'payload': b''}
```

I'm not getting any kind of payload in the response from get_status

**Reply by humbertogontijo** [reply]
Checkout to https://github.com/humbertogontijo/python-roborock/pull/18 branch and see if there is any improvement

**Reply by Lash-L** [reply]
Done - I end up getting a timeout on RoborockSocketListener connect.


            await self.loop.sock_connect(self.socket, (self.ip, 58867))

**Reply by Lash-L** [reply]
Ah ip was hardcoded in roborocklocalclient

I am now successfully getting status!

**Reply by Lash-L** [reply]
Willing to take any tasks on this you want me to, just let me know.

**Reply by humbertogontijo** [reply]
I can't figure out why get_map_v1 and get_multi_maps_list are not returning. Maybe you can figure it out playing with variables (I am trying other prefixes and protocols, but no success so far)

**Reply by Lash-L** [reply]
Okay I'll play around with it and see if I can figure anything out.

This is an aside, but I'm writing it down so I don't forget it/ to get any feedback from you.

Now that offline control is mainly possible, We should figure out the request for Product Info, then get the ip address and the mac address from that call.

By doing this, we can configure existing implementations to automatically add offline support without having to reconfigure.

Then new implementations can be discovered through dhcp in HA. And we can set the unique id of the device to the mac address

**Reply by Lash-L** [reply]
@humbertogontijo I started playing around with it - I've not used wireshark much(I've used mitm for previous reverse engineering), and I have a windows computer so I can't use airtool2.

When you're finding the important information in wireshark how do you do it? Do you filter to the robots ip address? or port 8883? or to udp? I know it theoretically needs to have 31 2e in it as that is hex for "1.0"

Any other tips you have for me so I can better try to help?

**Reply by humbertogontijo** [reply]
I filtered by src and target to the robot ip. 
Ignore the acknowledgment rows, focus on the ones with data. And then the payload is similar to the mqtt integration, but with a prefix of 4 bytes(which I'm not very familiar yet about the meaning).
The suffix after the payload is variable too(idk why). But ignoring what is after the payload Len + payload looked ok since there was no map data. 
Back to wire shark, copy the hex stream from the data and use the _decode_msg function to get the payload/protocol

**Reply by Lash-L** [reply]
Following you with all of that - but how do you deal with them needing to start with "1.0"?  I only get a small subset that look like they start with that and they all look identical, regardless of what I do in the app

**Reply by Lash-L** [reply]
Also do you take the hex stream and pass it directly into decode msg? or do you use bytes.fromhex()?

**Reply by humbertogontijo** [reply]
The first response have protocol 4 in it. There is no payload so ignore it and get the next one with protocol 5.
1.0 is like a flag telling the data started. Everything but the payload, id and timestamp will look the same

**Reply by humbertogontijo** [reply]
bytes.fromhex, yes

**Reply by Lash-L** [reply]
But the ones I am seeing dont' start with 1.0 and therefore fail in decode msg. I only see the exact same UDP request with the prefix different that start with 1.0 (and when I try to decode that it fails in the unpack stage)

![image](https://user-images.githubusercontent.com/20257911/229134669-ef005b39-5249-45b0-aaf1-a655bf3bff65.png)


This one appears to be mqtt data, but when I copy hex stream, bytes.fromhex() then run it through the decode_msg, it fails because it does not start with 1.0.encode()

I know I'm not being particularly helpful, but I think there is fundamentally something I am missing.

**Reply by humbertogontijo** [reply]
Its saying "Encrypted Application Data", that is probably why. Here is how mine looks like
<img width="1385" alt="image" src="https://user-images.githubusercontent.com/96875138/229138836-1839c284-9c3c-4572-9a1c-74a1dc98c4fc.png">

**Reply by Lash-L** [reply]
Mine looks like that for commands I send from offline.py.  Are you only looking at offline.py right now? Or are you able to look at the online packets as well?

**Reply by humbertogontijo** [reply]
The one from the photo is from airtool2 when I open the app in my phone

**Reply by Lash-L** [reply]
Okay I got an old mac, got airtool2 and wireshark on it. 
<img width="279" alt="Screen Shot 2023-03-31 at 12 01 14 PM" src="https://user-images.githubusercontent.com/20257911/229171706-9d41a673-536c-4316-8655-7c524cdb4edc.png">
Whenever I hit any of these captures, it opens wireshark, but only shows 802.11 protocol request. On my other wireshark instance I set it up to go through my wifi access point, should I do the same with airtool 2? How do you start yours typically?

**Reply by humbertogontijo** [reply]
You need to plug an iPhone to the Mac and go to "Capture iPhone Packet Trace"
<img width="294" alt="image" src="https://user-images.githubusercontent.com/96875138/229175826-69a1b80c-471e-491e-9353-6763b50d4ca1.png">

**Reply by Lash-L** [reply]
Ah -I see. that got it, thanks

**Reply by Lash-L** [reply]
Okay been playing around a bit - 

If you install pyshark, you can load the pcap file from wireshark directly into python and it makes it a bit easier to do some bulk experiments. 


```
capture = pyshark.FileCapture('wireshark.pcapng', 'rb')
for packet in capture:
    if hasattr(packet, "ip"):
        if packet.ip.dst == local_ip or packet.ip.src ==local_ip:
            if hasattr(packet, "DATA"):
                if hasattr(packet.DATA, "data"):
                    f = client._decode_msg(
                        bytes.fromhex(
                            packet.DATA.data),
                        local_key)
```

You end up getting a handful that have incorrect starts - which I just ignore. But then you also have a few that have an incorrect cr32 - including the call to get the ip/ mac.

Doesn't really help with the map issue, but I'm starting to get an understanding of everything

**Reply by Lash-L** [reply]
Okay - additional piece of info. It appears as if protocol 5 can sometimes have data - but its payload length is described at a different location? Not sure - you may have a better idea of how to figure this out. Right now, when it unpacks, it gets a payload len of 0. Even though there are dozens if not hundreds of extra bytes after :19.


Here's an example set of bytes that are protocol 5, and definitely have a payload(and fall other under criteria such as unencrypted tcp and has data.data

In the newest system, these will never run, as they have a payload length of 0. Any tips on how to go about reverse engineering this?

`b"1.0\x00\x00[\t\x00\x00\x1bXd'\x8d\xde\x00\x05\x00\x00\x00\x111.0\x00\x01/\xec\x00\x00\x1bYd'\x8d\xde\x00\x05\x00\x00\x00g1.0\x00\x00\x1d$\x00\x00\x1bZd'\x8d\xde\x00\x04\x00P%+\x1c \xe4D\x18\x11\xbc\xa6\xaa\xbe\xb0>\xe9\x10\xb7\x95u\xe6\xb8~\x84\xa0\x88\x935\xc9\xbc\xa2\xe7b7\x91Wr\x80\x10\xccY\xd2\xb3.\xf2\x8du\xf841\x00\xc9k\xad\xc1\x04\xb2\xc6\r\xc1\xa8Y\x18\xfc{X\x91\xf2\x06:\xf4\xe4\x83\x80\xfa=qE\xdfu\xf6\x98f\xa2\x12\x00\x00\x00g1.0\x00\x00\x1d%\x00\x00\x1b[d'\x8d\xde\x00\x04\x00P%+\x1c \xe4D\x18\x11\xbc\xa6\xaa\xbe\xb0>\xe9\x10\xb7\x95u\xe6\xb8~\x84\xa0\x88\x935\xc9\xbc\xa2\xe7b\xb2g\x91\xa3\xec\xa6\x19!\x14p\xfa>c\x85gTt\x8e\x8e\xf6g\xf6\xa8\xf9\xf0\xd4T\xa2\x0c\xad\xd0\xde\xa1n\x0c\x93i \xb1(\x01\x90\x93\x1c\r\xcd\rP\xc1\x0f\xdcC"`
